---
title: "hw2_geneticDistanceMethods_substitutionModels"
author: "Daisy Fry-Brumit"
date: "2022-09-19"
output:
  pdf_document: default
  html_document: default
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(msa)
library(phangorn)
library(phytools)
```
### Introduction

### Methods

  I collected whole mitochondrial genome access numbers for GenBank **citation** from Miya et. al and although there is some debate regarding the relationships among Lophiiformes and other ray-finned fishes, I used this same paper as a reference when assessing the performance of my tree building models. Whole genome access numbers were not all in use when I collected them from the paper, and I wanted to use individual loci within the genome, so I used the whole mitogenome access numbers as a jumping off point to narrow down the species I wanted to work with. Gene reference numbers for each locus and species can be found in **supplement1**.
  
  To quantify depth tolerance for each species, I consulted FishBase, an internationally maintained fish database hosted by Q-quatics **citation**. For each species, FishBase provides a range of depths individuals are found. Because I intended to consider depth tolerance, I selected the maximum known depth per species (**table1**).
  
Alignments were performed using MUSCLE in accordance with methods approached by Banon et. al in their paper on the efficacy of DNA barcoding in Lophiiformes **citation** 

Tree build were performed with both maximum likelihood (ML) and Bayesian methods using RAxML and MrBayes, respectively **citation** **citation**. Although the literature set a precedent for using ML was well as Bayesian methods, it should be noted that none of the nucleotide substitution model options offered by RAxML actually match best-fit models suggested by modelTest in R (**table2**). Models TIM1, TIM2, TIME3, TPM3u, and TrN were substituted for the GTR model in MrBayes following their parameter information **citation** and precedent set by Lecocq et. al **citation**

### Results

## evolutionary rates
all 13 loci failed the clockwise assumption

### Discussion

### References
Banon et. al
https://onlinelibrary.wiley.com/doi/full/10.1111/jzs.12281

parameter convention (https://evomics.org/resources/substitution-models/nucleotide-substitution-models/)

Lecocq et. al
https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0065642
